File: 0938_range_sum_of_BST/range_sum_BST.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    
    int res = 0;
    int low;
    int high;
    
    // Time: O(n)
    // Space: O(n)
    public int rangeSumBST(TreeNode root, int low, int high) {
        this.low = low;
        this.high = high;
        dfs(root);
        return res;
    }
    
    private void dfs(TreeNode node) {
        if (node == null) {
            return;
        }
        if (node.val >= low && node.val <= high) {
            res += node.val;
        }
        if (node.val > low) {
            dfs(node.left);
        }
        if (node.val < high) {
            dfs(node.right);
        }
    }
}

File: 0938_range_sum_of_BST/range_sum_BST.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def rangeSumBST(root: TreeNode, low: int, high: int) -> int:
    """
    Time: O(n)
    Space: O(n)
    """
    res = 0
    def helper(node: TreeNode):
        nonlocal res
        if not node:
            return
        if high >= node.val >= low:
            res += node.val
        if node.val > low:
            helper(node.left)
        if node.val < high:
            helper(node.right)
    helper(root)
    return res